src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs

```diff
@@ -1174,21 +1182,36 @@ internal ImmutableArray<LocalInfo<TypeSymbol>> GetLocalInfo(StandaloneSignatureH
         }
 
         /// <exception cref="UnsupportedSignatureContent">If the encoded parameter type is invalid.</exception>
-        private void DecodeParameterOrThrow(ref BlobReader signatureReader, /*out*/ ref ParamInfo<TypeSymbol> info)
+        private void DecodeParameterOrThrow(ref BlobReader signatureReader, /*out*/ ref ParamInfo<TypeSymbol> info, bool isReturn)
         {
+            var allowedRequiredModifiers = AllowedRequiredModifierType.System_Runtime_InteropServices_InAttribute;
+            if (isReturn)
+            {
+                // PROTOTYPE(init-only): can we make this more restrictive (ie. disallow aside from the return value of an instance setter)?
+                allowedRequiredModifiers |= AllowedRequiredModifierType.System_Runtime_CompilerServices_IsExternalInit;
+            }
+
             info.CustomModifiers = DecodeModifiersOrThrow(
                 ref signatureReader,
-                AllowedRequiredModifierType.System_Runtime_InteropServices_InAttribute,
+                allowedRequiredModifiers,
                 out SignatureTypeCode typeCode,
-                out bool inAttributeFound);
+                out AllowedRequiredModifierType inAttributeFound);
 
             if (typeCode == SignatureTypeCode.ByReference)
             {
                 info.IsByRef = true;
+
+                // PROTOTYPE(init-only): we're incorrectly letting modreq(IsExternalInit) into the RefCustomModifiers
```

Comment:
>// PROTOTYPE(init-only): we're incorrectly letting modreq(IsExternalInit) into the RefCustomModifiers [](start = 16, length = 101)

It looks like this comment is obsolete, the ``if``` at the end of this block takes care of the issue. BTW, consider moving that ```if``` here.

Reply:
I'll remove the comment. Thanks
It's a nit, but the reason the `if/throw` is at the end of this block is that the decoded modreq will get stored rather than lost.

---
In reply to: [417491981](https://github.com/dotnet/roslyn/pull/43275#discussion_r417491981) [](ancestors = 417491981)

Reply:
> the reason the if/throw is at the end of this block is that the decoded modreq will get stored rather than lost.

This isn't really useful. I prefer us to follow consistent handling of unsupported modereqs within this type - throw as soon as we recognize the fact

---
In reply to: [417569537](https://github.com/dotnet/roslyn/pull/43275#discussion_r417569537) [](ancestors = 417569537,417491981)

Reply:
It might make sense to do that after the assignment```info.RefCustomModifiers = info.CustomModifiers```, but it is really not worth trying to decode the next set of modifiers because we aren't even going to try decoding the type of the parameter.

---
In reply to: [417585593](https://github.com/dotnet/roslyn/pull/43275#discussion_r417585593) [](ancestors = 417585593,417569537,417491981)